dask/array/core.py

```diff
@@ -3587,6 +3597,90 @@ def _vindex_array(x, dict_indexes):
     return result_1d.reshape(broadcast_shape + result_1d.shape[1:])
 
 
+def _vindex_dask_arrays(x, dict_indexes):
+    """ Point wise indexing with Dask Arrays
+
+    Makes two passes of the data. First pass with all static values: integers,
+    slices, lists, and NumPy Arrays. Second pass (done in this function)
+    handles Dask Arrays using ``atop``. The strategy there involves comparing
+    Dask Array indices with ``arange`` in chunks to determine the relative
+    indexing needed to extract values for that chunk. This strategy works just
+    as well for Dask Arrays of known and unknown length.
+    """
+
+    from .creation import arange
+
+    # Construct a tuple of indices for selection.
+    # Also collect axes that are part of the result.
+    #
+    # * arr_inds -- used for second pass
+    # * out_axes -- track where each
+    #
+    arr_inds = []
+    out_axes = []
+    for i in range(x.ndim):
+        try:
+            ind = dict_indexes[i]
+        except KeyError:
+            arr_inds.append(slice(None))
+            out_axes.append(len(arr_inds))
+        else:
+            arr_inds.append(ind)
+            if 0 not in out_axes:
+                out_axes.append(0)
+    arr_inds = tuple(arr_inds)
+    out_axes = tuple(out_axes)
+
+    # Find indices for all points in x to aid in selection within chunks
+    x_inds = tuple(
+        arange(s, chunks=c) for s, c in zip(x.shape, x.chunks)
+    )
+
+    # Use global indices for reference in applying selected indices relative
+    # to each chunk. For static values applied in the first pass, pass along
+    # a dummy scalar value to notify the chunk should be kept unchanged.
+    idx_vals = atop(
+        _vindex_dask_arrays_chunk,
+        out_axes,
+        x, tuple(range(1, 1 + x.ndim)),
+        *concat([
+            (e_0, (0,), e_i, (i,)) if isinstance(e_0, Array) else
+            (e_0, None, e_i, (i,))
+            for i, (e_0, e_i) in enumerate(zip(arr_inds, x_inds), start=1)
+        ]),
+        dtype=x.dtype,
+        concatenate=True
```

Comment:
Something to watch out for: this operation is lazy, but isn't really distributed friendly. If I understand correctly, `concatenate=True` means that the indexing arrays essentially get concatenated together across chunks. So things will break if your indexing arrays are too big to fit into memory.

This could conceivably be avoided, but it would require some tricky logic to merge across chunks.

Reply:
Alright, will give this some thought. Thanks.

Reply:
AFAICT `concatenate=True` is just a convenience option that gives one a single NumPy array instead of some list of NumPy arrays that `atop` would otherwise provide to the kernel function. ( https://github.com/dask/dask/pull/1609 ) The amount of data given in each case (`concatenate=False`/`concatenate=True`) is the same. Historically it sounds like that wasn't always true.

There may be some argument to not performing the concatenation as the memory usage doubles in the process (at least temporarily). That said, one normally chooses chunk sizes knowing that copies may be made (maybe even a few). So this doesn't seem unreasonable to me.

Reply:
`concatenate=True` hasn't changed, and indeed `atop` has always processed multiple inputs at once when output indices are not available. But loading all input chunks corresponding to dropped outputs can still be problematic.

Looking at this again, I think you actually consolidate all *indexed* chunks along these dimensions, e.g., with `array.vindex[key]`, then `array` could potentially get computed. This would be problematic in the case where `array` corresponds to a large tiled image, e.g., a 100k x 100k matrix with chunks of size 2000 x 2000.

That said, I don't think dask (without distributed) is capable in handling this in any sane way, since it really requires dynamic task creation. Otherwise you're stuck with all combinations of indexed and indexer chunks, which will quickly get overwhelming for moderately sized numbers of chunks.